File: src/edu/uf/interactable/Lactoferrin.java
package edu.uf.interactable;

import java.util.HashMap;
import java.util.Map;

import edu.uf.Diffusion.Diffuse;
import edu.uf.compartments.GridFactory;
import edu.uf.intracellularState.Phenotype;
import edu.uf.primitives.Interactions;
import edu.uf.utils.Constants;
import edu.uf.utils.Util;

public class Lactoferrin extends Molecule{
    

	public static final String NAME = "Lactoferrin";
	public static final int NUM_STATES = 3;
    private static final double THRESHOLD = Constants.K_M_TF_LAC * Constants.VOXEL_VOL / 1.0e6;
    
    private static Lactoferrin molecule = null; 

    private static final Map<String, Integer> INDEXES;
    
    static {
    	INDEXES = new HashMap<>();
    	INDEXES.put("Lactoferrin", 0);
    	INDEXES.put("LactoferrinFe", 1);
    	INDEXES.put("LactoferrinFe2", 2);
    }
    
    private Lactoferrin(double[][][][] qttys, Diffuse diffuse) {
		super(qttys, diffuse);
		this.setPhenotye(Phenotype.createPhenotype());
	}
    
    public static Lactoferrin getMolecule(Diffuse diffuse) {
    	if(molecule == null) {
    		double[][][][] values = new double[NUM_STATES][GridFactory.getXbin()][GridFactory.getYbin()][GridFactory.getZbin()];
    		molecule = new Lactoferrin(values, diffuse); 
    	}
    	return molecule;
    }
    
    public static Lactoferrin getMolecule() {
    	return getMolecule(null);
    }
    
    @Override
    public double getKd() {
    	return -1;
    }
    
    public void degrade() {} //REVIEW

    public int getIndex(String str) {
        return Lactoferrin.INDEXES.get(str);
    }

    public void computeTotalMolecule(int x, int y, int z) {
    	this.totalMoleculesAux[0] = this.totalMoleculesAux[0] + this.get(0, x, y, z);
    	this.totalMoleculesAux[1] = this.totalMoleculesAux[1] + this.get(1, x, y, z);
    	this.totalMoleculesAux[2] = this.totalMoleculesAux[2] + this.get(2, x, y, z);
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        if (interactable instanceof Macrophage)
        	return Interactions.lactoferrinMacrophageUpatake((Macrophage) interactable, this, x, y, z);
        	
        if (interactable instanceof Neutrophil) 
        	return Interactions.lactoferrinDegranulation((Neutrophil) interactable, this, x, y, z);
        
        if (interactable instanceof Transferrin) 
           return Interactions.lactoferrinTransferrinChelation((Molecule) interactable, this, x, y, z);
        
        if (interactable instanceof Iron) 
        	return Interactions.transferrinIronChelation(this, (Molecule) interactable, x, y, z); //this method can be used for both lactoferrin and transferrin.
        
        return interactable.interact(this, x, y, z);
    }

	@Override
	public String getName() {
		return NAME;
	}

	@Override
	public double getThreshold() {
		return THRESHOLD;
	}

	@Override
	public int getNumState() {
		return NUM_STATES;
	}
	
	@Override
	public boolean isTime() {
		return true;
	}
        
}

File: src/edu/uf/interactable/Lactoferrin.cpp
/*
 * Lactoferrin.cpp
 *
 *  Created on: Jul 19, 2024
 *      Author: henriquedeassis
 */

#include "Lactoferrin.h"
#include "../utils/Util.h"
#include "../utils/Id.h"
#include "../primitives/Interactions.h"
#include "Macrophage.h"
#include "Neutrophil.h"
#include "Transferrin.h"
#include <memory>

namespace edu {
namespace uf {
namespace interactable {

using namespace primitives;

const int Lactoferrin::classId = Id::getId();
const string Lactoferrin::NAME = "Lactoferrin";
const double Lactoferrin::THRESHOLD = utils::constexprants::K_M_TF_LAC * utils::constexprants::VOXEL_VOL / 1.0e6;
Lactoferrin* Lactoferrin::molecule = nullptr;

Lactoferrin::Lactoferrin(double**** qttys, int xbin, int ybin, int zbin, char wbin, Diffuse* diffuse)
    : Molecule(qttys, xbin, ybin, zbin, wbin, diffuse) {}

Lactoferrin* Lactoferrin::getMolecule(Diffuse* diffuse, int xbin, int ybin, int zbin, char wbin) {
	if (molecule == nullptr) {
	        double**** values = new double***[wbin];
	        for (int i = 0; i < wbin; ++i) {
	            values[i] = new double**[xbin];
	            for (int j = 0; j < xbin; ++j) {
	                values[i][j] = new double*[ybin];
	                for (int k = 0; k < ybin; ++k) {
	                    values[i][j][k] = new double[zbin]();
	                }
	            }
	        }
	        molecule = new Lactoferrin(values, xbin, ybin, zbin, wbin, diffuse);
	    }
	    return molecule;
}

int Lactoferrin::getClassId() const {
	return classId;
}

double Lactoferrin::getKd() {
    return -1;
}

int Lactoferrin::getIndex(const char* str) const{
	return INDEXES.at(str);
}

void Lactoferrin::degrade() {}

void Lactoferrin::computeTotalMolecule(int x, int y, int z) {
	totalMoleculesAux[0] += get(0, x, y, z);
	totalMoleculesAux[1] += get(1, x, y, z);
	totalMoleculesAux[2] += get(2, x, y, z);
}

bool Lactoferrin::templateInteract(Interactable* interactable, int x, int y, int z) {
    if (auto macrophage = dynamic_cast<Macrophage*>(interactable))
        return Interactions::lactoferrinMacrophageUpatake(macrophage, this, x, y, z);

    if (auto neutrophil = dynamic_cast<Neutrophil*>(interactable))
        return Interactions::lactoferrinDegranulation(neutrophil, this, x, y, z);

    if (auto transferrin = dynamic_cast<Transferrin*>(interactable))
    	return Interactions::lactoferrinTransferrinChelation(transferrin, this, x, y, z);

    return interactable->interact(this, x, y, z);
}

string Lactoferrin::getName() const {
    return NAME;
}

double Lactoferrin::getThreshold() const {
    return THRESHOLD;
}

int Lactoferrin::getNumState() const {
    return NUM_STATES;
}

bool Lactoferrin::isTime() const {
    return true;
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */
